File: sdk-ts/test/test_tools_json.ts
/**
 * Validates keytools/tools.json: schema, unique URLs, and per-tool keyword
 * round-trip (same logic as findToolsLocal). Prints one line per tool — like a linter.
 *
 * Usage:
 *   npx ts-node test/test_tools_json.ts
 *   npx ts-node test/test_tools_json.ts --quiet   # only failures + summary
 */

import * as fs from "fs";
import * as path from "path";
import { matchPromptAgainstKeywords } from "../src/find_tools_local";

const DB_PATH = path.resolve(__dirname, "../../keytools/tools.json");
const QUIET = process.argv.includes("--quiet");

interface RawTool {
  url: string;
  title: string;
  description: string;
  image: { large: string | null; small: string | null };
  rules: { keywords: string[] };
  texts?: unknown;
}

function toolSlug(url: string): string {
  try {
    const last = new URL(url).pathname.split("/").filter(Boolean).pop();
    return last ?? url;
  } catch {
    return url;
  }
}

function shortTitle(title: string, max = 72): string {
  const s = title.trim();
  return s.length <= max ? s : `${s.slice(0, max - 1)}…`;
}

function isHttps(s: string): boolean {
  return /^https:\/\//i.test(s.trim());
}

function validateToolsJson(raw: unknown): string[] {
  const errors: string[] = [];
  if (!raw || typeof raw !== "object" || !("tools" in raw)) {
    return ["root must be an object with a tools array"];
  }
  const tools = (raw as { tools: unknown }).tools;
  if (!Array.isArray(tools) || tools.length === 0) {
    return ["tools must be a non-empty array"];
  }

  const seenUrls = new Set<string>();

  tools.forEach((t, i) => {
    const p = `tools[${i}]`;
    if (!t || typeof t !== "object") {
      errors.push(`${p}: must be an object`);
      return;
    }
    const o = t as Record<string, unknown>;

    for (const key of ["url", "title", "description", "image", "rules"] as const) {
      if (!(key in o)) {
        errors.push(`${p}: missing "${key}"`);
      }
    }
    if (errors.some((e) => e.startsWith(p + ":") && e.includes("missing"))) {
      return;
    }

    const url = o.url;
    const title = o.title;
    const description = o.description;
    const image = o.image;
    const rules = o.rules;

    if (typeof url !== "string" || !url.trim()) {
      errors.push(`${p}.url: non-empty string required`);
    } else if (!isHttps(url)) {
      errors.push(`${p}.url: must be an https URL`);
    } else if (seenUrls.has(url)) {
      errors.push(`${p}.url: duplicate URL`);
    } else {
      seenUrls.add(url);
    }

    if (typeof title !== "string" || !title.trim()) {
      errors.push(`${p}.title: non-empty string required`);
    }
    if (typeof description !== "string" || !description.trim()) {
      errors.push(`${p}.description: non-empty string required`);
    }

    if (!image || typeof image !== "object") {
      errors.push(`${p}.image: object required`);
    } else {
      const im = image as Record<string, unknown>;
      for (const sz of ["large", "small"] as const) {
        const u = im[sz];
        if (u === null) continue;
        if (typeof u !== "string" || !u.trim()) {
          errors.push(
            `${p}.image.${sz}: null or non-empty https URL string required`
          );
        } else if (!isHttps(u)) {
          errors.push(`${p}.image.${sz}: must be an https URL`);
        }
      }
    }

    if (!rules || typeof rules !== "object") {
      errors.push(`${p}.rules: object required`);
    } else {
      const kws = (rules as { keywords?: unknown }).keywords;
      if (!Array.isArray(kws) || kws.length === 0) {
        errors.push(`${p}.rules.keywords: non-empty array required`);
      } else {
        kws.forEach((kw, j) => {
          if (typeof kw !== "string" || !kw.trim()) {
            errors.push(`${p}.rules.keywords[${j}]: non-empty string required`);
          }
        });
      }
    }

    const texts = o.texts;
    if (texts !== undefined) {
      if (!texts || typeof texts !== "object") {
        errors.push(`${p}.texts: must be an object when present`);
      } else {
        const ru = (texts as { ru?: unknown }).ru;
        if (ru !== undefined) {
          if (!ru || typeof ru !== "object") {
            errors.push(`${p}.texts.ru: must be an object when present`);
          } else {
            const r = ru as Record<string, unknown>;
            for (const fld of ["title", "description"] as const) {
              if (typeof r[fld] !== "string" || !(r[fld] as string).trim()) {
                errors.push(`${p}.texts.ru.${fld}: non-empty string required`);
              }
            }
          }
        }
      }
    }
  });

  return errors;
}

function run(): void {
  const rawText = fs.readFileSync(DB_PATH, "utf-8");
  let data: unknown;
  try {
    data = JSON.parse(rawText);
  } catch (e) {
    console.error("[FAIL] tools.json is not valid JSON:", e);
    process.exit(1);
  }

  if (!QUIET) {
    console.log(`\n>>> ${DB_PATH}`);
    console.log(">>> Schema");
  }

  const schemaErrors = validateToolsJson(data);
  if (schemaErrors.length > 0) {
    console.error("[FAIL] Schema validation:");
    for (const line of schemaErrors) {
      console.error(`       ${line}`);
    }
    process.exit(1);
  }

  const tools = (data as { tools: RawTool[] }).tools;
  if (!QUIET) {
    console.log(`[PASS] schema — ${tools.length} entries\n`);
    console.log(">>> Per-tool (keywords → synthetic prompt → match)");
  }

  let passed = 0;
  let failed = 0;

  for (let i = 0; i < tools.length; i++) {
    const tool = tools[i];
    const kws = tool.rules.keywords;
    const prompt = kws.join(" ");
    const slug = toolSlug(tool.url);
    const label = `[${i}] ${slug}  ${shortTitle(tool.title)}`;

    if (matchPromptAgainstKeywords(prompt, kws)) {
      passed++;
      if (!QUIET) {
        console.log(`[PASS] ${label}`);
      }
    } else {
      failed++;
      console.error(`[FAIL] ${label}`);
      console.error(`       keywords do not match synthetic prompt`);
    }
  }

  console.log(
    `\n${failed === 0 ? "[PASS]" : "[FAIL]"} tools.json — ${passed}/${tools.length} per-tool checks OK, schema OK`
  );

  process.exit(failed > 0 ? 1 : 0);
}

run();

File: sdk-py/test/test_tools_json.py
#!/usr/bin/env python3
"""
Validates keytools/tools.json: schema, unique URLs, and per-tool keyword
round-trip (same logic as find_tools_local). Prints one line per tool — like a linter.

Usage:
  python test/test_tools_json.py
  python test/test_tools_json.py --quiet   # only failures + summary
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from urllib.parse import urlparse

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))
from find_tools_local import match_prompt_against_keywords

DB_PATH = Path(__file__).parent.parent.parent / "keytools" / "tools.json"
_HTTPS = re.compile(r"^https://", re.I)


def _quiet(argv: list[str]) -> bool:
    return "--quiet" in argv


def _tool_slug(url: str) -> str:
    try:
        p = urlparse(url).path.strip("/").split("/")
        return p[-1] if p else url
    except Exception:
        return url


def _short_title(title: str, max_len: int = 72) -> str:
    s = title.strip()
    return s if len(s) <= max_len else s[: max_len - 1] + "…"


def _is_https(s: str) -> bool:
    return bool(_HTTPS.match(s.strip()))


def validate_tools_json(raw: object) -> list[str]:
    errors: list[str] = []
    if not isinstance(raw, dict) or "tools" not in raw:
        return ["root must be an object with a tools array"]
    tools = raw["tools"]
    if not isinstance(tools, list) or len(tools) == 0:
        return ["tools must be a non-empty array"]

    seen_urls: set[str] = set()

    for i, t in enumerate(tools):
        p = f"tools[{i}]"
        if not isinstance(t, dict):
            errors.append(f"{p}: must be an object")
            continue

        for key in ("url", "title", "description", "image", "rules"):
            if key not in t:
                errors.append(f'{p}: missing "{key}"')

        if any(e.startswith(f"{p}:") and "missing" in e for e in errors):
            continue

        url = t["url"]
        title = t["title"]
        description = t["description"]
        image = t["image"]
        rules = t["rules"]

        if not isinstance(url, str) or not url.strip():
            errors.append(f"{p}.url: non-empty string required")
        elif not _is_https(url):
            errors.append(f"{p}.url: must be an https URL")
        elif url in seen_urls:
            errors.append(f"{p}.url: duplicate URL")
        else:
            seen_urls.add(url)

        if not isinstance(title, str) or not title.strip():
            errors.append(f"{p}.title: non-empty string required")
        if not isinstance(description, str) or not description.strip():
            errors.append(f"{p}.description: non-empty string required")

        if not isinstance(image, dict):
            errors.append(f"{p}.image: object required")
        else:
            for sz in ("large", "small"):
                u = image.get(sz)
                if u is None:
                    continue
                if not isinstance(u, str) or not u.strip():
                    errors.append(
                        f"{p}.image.{sz}: null or non-empty https URL string required"
                    )
                elif not _is_https(u):
                    errors.append(f"{p}.image.{sz}: must be an https URL")

        if not isinstance(rules, dict):
            errors.append(f"{p}.rules: object required")
        else:
            kws = rules.get("keywords")
            if not isinstance(kws, list) or len(kws) == 0:
                errors.append(f"{p}.rules.keywords: non-empty array required")
            else:
                for j, kw in enumerate(kws):
                    if not isinstance(kw, str) or not kw.strip():
                        errors.append(
                            f"{p}.rules.keywords[{j}]: non-empty string required"
                        )

    return errors


def run(argv: list[str]) -> int:
    quiet = _quiet(argv)

    try:
        raw_text = DB_PATH.read_text(encoding="utf-8")
    except OSError as e:
        print(f"[FAIL] cannot read {DB_PATH}: {e}", file=sys.stderr)
        return 1

    try:
        data = json.loads(raw_text)
    except json.JSONDecodeError as e:
        print(f"[FAIL] tools.json is not valid JSON: {e}", file=sys.stderr)
        return 1

    if not quiet:
        print(f"\n>>> {DB_PATH}")
        print(">>> Schema")

    schema_errors = validate_tools_json(data)
    if schema_errors:
        print("[FAIL] Schema validation:", file=sys.stderr)
        for line in schema_errors:
            print(f"       {line}", file=sys.stderr)
        return 1

    tools = data["tools"]
    if not quiet:
        print(f"[PASS] schema — {len(tools)} entries\n")
        print(">>> Per-tool (keywords → synthetic prompt → match)")

    passed = 0
    failed = 0

    for i, tool in enumerate(tools):
        kws = tool["rules"]["keywords"]
        prompt = " ".join(kws)
        slug = _tool_slug(tool["url"])
        label = f'[{i}] {slug}  {_short_title(tool["title"])}'

        if match_prompt_against_keywords(prompt, kws):
            passed += 1
            if not quiet:
                print(f"[PASS] {label}")
        else:
            failed += 1
            print(f"[FAIL] {label}", file=sys.stderr)
            print("       keywords do not match synthetic prompt", file=sys.stderr)

    status = "[PASS]" if failed == 0 else "[FAIL]"
    print(
        f"\n{status} tools.json — {passed}/{len(tools)} per-tool checks OK, schema OK"
    )
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(run(sys.argv[1:]))
